posts/2021-02-11-Rust-1.50.0.md

```diff
@@ -0,0 +1,140 @@
+release: true
+---
+
+The Rust team is happy to announce a new version of Rust, 1.50.0. Rust is a
+programming language that is empowering everyone to build reliable and
+efficient software.
+
+If you have a previous version of Rust installed via rustup, getting Rust
+1.50.0 is as easy as:
+
+```console
+rustup update stable
+```
+
+If you don't have it already, you can [get `rustup`][install]
+from the appropriate page on our website, and check out the
+[detailed release notes for 1.50.0][notes] on GitHub.
+
+[install]: https://www.rust-lang.org/install.html
+[notes]: https://github.com/rust-lang/rust/blob/master/RELEASES.md#version-1500-2021-02-11
+
+## What's in 1.50.0 stable
+
+For this release, we have improved array indexing, expanded union field safety, refined file descriptors, and new library additions.
+See the [detailed release notes][notes] to learn about other changes
+not covered by this post.
+
+### Const-generic array indexing
+
+Continuing the march toward stable `const` generics, this release adds
+implementations of `ops::Index` and `IndexMut` for arrays `[T; N]` for
+_any_ length of `const N`. The indexing operator `[]` already worked on
+arrays through built-in compiler magic, but at the type level, arrays
+didn't actually implement the library traits until now.
+
+```rust
+fn second<C>(container: &C) -> &C::Output
+where
+    C: std::ops::Index<usize> + ?Sized,
+{
+    &container[1]
+}
+
+fn main() {
+    let array: [i32; 3] = [1, 2, 3];
+    assert_eq!(second(&array[..]), &2); // slices worked before
+    assert_eq!(second(&array), &2); // now it also works directly
+}
+```
+
+### Safe assignments to `ManuallyDrop<T>` union fields
+
+Rust 1.49 made it possible to add `ManuallyDrop<T>` fields to a `union`
+as part of allowing `Drop` for unions at all; however, unions don't know
+which variant is valid (they don't drop any old value on assignment), so
+safe Rust has limited this to `Copy` types only, which never `Drop`. Of
+course, that's also true of `ManuallyDrop`, so now Rust 1.50 allows safe
+assignments to these fields as well.
+
+### A niche for `File` on Unix platforms
+
+Some types in Rust have specific limitations on what is considered
+valid value, which may not cover the entire range of possible memory
+values. We call any remaining invalid value a [niche], and this space
+may be used for type layout optimizations. For example, in Rust 1.28
+we introduced `NonZero` types where `0` is a niche, and this allowed
+`Option<NonZero>` to use `0` to represent `None` with no extra memory.
+
+On Unix platforms, Rust's `File` is simply made of the system's integer
+file descriptor, and this happens to have a possible niche
+as well because it can never be `-1`! System calls which return a file
+descriptor use `-1` to indicate that an error occurred (check `errno`)
+so it's never possible for `-1` to be a real file descriptor. Starting
+in Rust 1.50 this niche is represented to the compiler so it can be
```

Comment:
"represented" sounds a bit funny to me. However, I can't think of a better word to use ;)

Does someone have an idea of a better word? If not, it's probably fine as-is though.

Reply:
how about "this niche is added to Rust's type system so it can be used in layout optimizations too"

Reply:
IIUC, it's not that it's added to the type system – that implies it was hard-coded. It's more like it's *exposed* to the type system, so maybe this?

```suggestion
in Rust 1.50 information about this niche is exposed to the compiler so it can be
```

Reply:
how about "this niche is added to the type's definition so it can be used in layout optimizations too"

Reply:
since the niche didn't exist before